lib/lumberjack/server.rb

```diff
@@ -42,31 +42,42 @@ def initialize(options={})
       @ssl.key = OpenSSL::PKey::RSA.new(File.read(@options[:ssl_key]),
                                         @options[:ssl_key_passphrase])
       @ssl_server = OpenSSL::SSL::SSLServer.new(@tcp_server, @ssl)
+      @stop = false
     end # def initialize
 
     def run(&block)
-      while true
-        connection = accept
-        Thread.new(connection) do |connection|
+      while !@stop
+        sock = @tcp_server.accept
+        Thread.new(sock) do |sock|
+          connection = accept_ssl(sock)
           connection.run(&block)
         end
       end
     end # def run
 
-    def accept(&block)
+    def accept_ssl(socket)
+      tries = 0
       begin
-        fd = @ssl_server.accept
+        ssl = OpenSSL::SSL::SSLSocket.new(socket, @ssl)
+        ssl.sync_close = true
+        Connection.new(ssl.accept)
       rescue EOFError, OpenSSL::SSL::SSLError, IOError
         # ssl handshake or other accept-related failure.
         # TODO(sissel): Make it possible to log this.
-        retry
-      end
-      if block_given?
-        block.call(fd)
-      else
-        Connection.new(fd)
+        if tries <= 3
```

Comment:
Best to abort on all errors I think unless they are explicitly known to be temporary.